packages/shared/src/authz/domain/loaders.ts

```diff
@@ -8,11 +10,47 @@ import type {
   GetWorkspaceSsoSession
 } from './workspaces/operations.js'
 
-export type AuthCheckContext<LoaderKeys extends keyof AuthCheckContextLoaders> = {
-  loaders: Pick<AuthCheckContextLoaders, LoaderKeys>
+// utility type that ensures all properties functions that return promises
+type PromiseAll<T> = {
+  [K in keyof T]: T[K] extends (...args: infer Args) => MaybeAsync<infer Return>
+    ? (...args: Args) => Promise<Return>
+    : never
 }
 
-export type AuthCheckContextLoaders = {
+// wrapper type for AllAuthCheckContextLoaders that ensures loaders follow the expected schema
+type AuthContextLoaderMappingDefinition<
+  Mapping extends {
+    [Key in keyof Mapping]: Key extends AuthCheckContextLoaderKeys
+      ? Mapping[Key]
+      : never
+  }
+> = PromiseAll<
+  OverrideProperties<
+    {
+      [key in AuthCheckContextLoaderKeys]: unknown
+    },
+    Mapping
+  >
+>
+
+/**
+ * All loaders must be listed here for app startup validation to work properly
+ */
+export const AuthCheckContextLoaderKeys = <const>{
```

Comment:
1. didn't really make sense that a runtime collection of the expected loader keys lived in `server` not `shared`, so i moved it over
2. also made it an enum so that (optionally) loaders can be referenced through the enum without string literals